### ISSUE TITLE
[FEATURE] Support for reducers (e.g., MapReduce)

### ISSUE THREAD

**Issue by skyzyx** [hashicorp/terraform#27716]
### Current Terraform Version

0.14.6

### Use-cases

User-provided content (known at "runtime") NOT known ahead of time by the module (analogous to "compile time"). Being able to process data like the _reduce_ part of MapReduce where each new step in the iteration performs actions against the results of a previous step in the iteration.

Much like `for` or `for_each`, but `iteration[1]` is able to use the _results_ of `iteration[0]` as its input. Or perhaps, the entire preceding stack of results would be available. Think: performing the fibonacci sequence to 10 places using HCL+Terraform and a new reducer construct (e.g., `reduce()`).

Specifically, I'd like to perform multiple, sequential regex replacements on a string which may or may not overlap.

### Proposal

I was originally thinking something like:

```hcl
{ reduce i, v, stack in range(10): try(stack[length(stack)-2], 0) + try(stack[length(stack)-1], i) }
```

I'm looking at the example at https://www.terraform.io/docs/language/functions/range.html. Something like that _could_ work, provided that the output result is accessible to the reducer _while_ it's running.

**Comment by nl-brett-stime** [comment]
My use case:

Take a list of strings and remove them as substrings from another string.

E.g.,

```
locals {
  words_to_remove = ["foo", "bar", "baz"]
  string_to_clean = "somebazfoobarstring"
  
  # Results in "somestring"
  cleaned_string = reduce(, words_to_remove, (acc, next) => substr(acc, next, ""))
}
```

It seems like this would require lambdas / anonymous functions, though. Maybe it'd be easier from a language-design standpoint to do it as a reduce comprehension / expression (similar to for-expressions)?

```
locals {
  
  cleaned_string2 = [reduce acc, next from concat([local.string_to_clean], local.words_to_remove) => substr(acc, next, "")][0]
}
```

...or maybe some other braces to indicate that the expression might reduce down to a scalar or string (as opposed to a collection)?
```
locals {
  
  cleaned_string2 = <reduce acc, next from concat([local.string_to_clean], local.words_to_remove) => substr(acc, next, "")>
}
```

**Comment by skyzyx** [comment]
It just occurred to me — could something like this be implemented as a Provider in Go?

Coincidentally, I've been working with the `hcl/v2/hclparse` and `go-cty/cty` packages over the last couple of days, and while those libraries are short on examples (especially `cty.DynamicVal`), I'm wondering if this could be constructed as a `data` source via a custom provider (assuming that the core team doesn't have bandwidth to work on this feature since I requested it over a year ago).

I don't know that there's way to expose new _functions_ to user-land Terraform files, but perhaps some input/output magic? Similar to [what I proposed here](https://github.com/hashicorp/terraform/issues/27696#issuecomment-877437227)? Or should we be looking toward to Go support in the [Terraform CDK](https://www.terraform.io/cdktf) for that kind of functionality?

Thoughts, @apparentlymart, @jbardin, or other committers?

**Comment by apparentlymart** [comment]
The example syntax shown in the original writeup would only be implementable via an entirely new expression type in HCL itself. Unfortunately I think blocking something like that would probably be blocked by backward compatibility, because it's currently valid to write e.g. `{ reduce = "a" }` to define an object with an attribute named `reduce`, and I believe it would require unbounded lookahead to distinguish that from a "`reduce` expression". That ambiguity is already present for `for` expressions but has been there since Terraform v0.12 and so `{ for = "a" }` has been invalid for a long time. (You have to write `{ "for" = "a" }` if you want to write an object expression like that, since the interpretation of `for` as a keyword takes precedence to resolve the ambiguity.)

With that said, I might argue that `{ .. }` is a quirky choice of punctuation for this situation anyway, since the result of `reduce` is not an object as would be the case for `{ for ... }`. A `reduce` function can return a value of any type, so I guess we'd probably want to build it around parentheses as the bracketing delimiter to help indicate that, but of course there is still ambiguity in that path which we'd need to resolve somehow:

```hcl
( reduce i, v, stack in range(10): try(stack[length(stack)-2], 0) + try(stack[length(stack)-1], i) )
```

---

I think this would be pretty challenging to implement as a built-in function, too. We could in principle use the special building blocks that allow the `try` and `can` functions to evaluate their expressions in a special way, but I think the closest we could get to the opening example with that would be something like this:

```hcl
reduce([i, v, stack], range(10), try(stack[length(stack)-2], 0) + try(stack[length(stack)-1], i))
```

To achieve this we'd need to specify `reduce` so that:
* The first and third arguments are annotated in the same way as the `try` arguments, so that the function implementation gets passed the expression itself rather than the result of evaluating it. For the first argument the naked expression would be sufficient, while the third argument would need to capture a closure so that the expression can potentially refer to variables other than `i`, `v`, and `stack`.
* For the first argument, we'd need to do [static analysis](https://github.com/hashicorp/hcl/blob/main/spec.md#static-analysis) on it, rather than normal expression evaluation, in order to treat it as a list of variable declarations rather than an value. Specifically, we'd first use "Static List" to get a list of the expressions inside the brackets, and then on each one "Static Traversal" to get access to the literal names `"i"`, `"v"`, and `"stack"`.
* The second argument could just evaluate as normal, I think, to determine the list of inputs to reduce.
* The third argument would need custom evaluation where we create a nested scope under the scope in the captured closure which contains the `i`, `v`, and `stack` variables. That expression would get evaluated one for each reduce step, with the final call becoming the result.

Although it does seem technically possible to implement something like that, we don't have any existing precedent for something shaped this way, and so that makes me reluctant to introduce it: subjectively, I don't think it's particularly clear to an unfamiliar reader what this means and I would expect to be constantly answering questions about it from those who have inherited existing modules that make use of it.

---

You also asked about whether it could be implemented in a provider. Providers don't have any access to evaluate expressions themselves, because by the time they are running we've already done all of the evaluation. Even if we wanted to allow it, it would be quite challenging since there is a wire protocol between Terraform Core and a provider and we have no existing mechanism to serialize an expression closure to send it over a wire protocol to be evaluated at the other end.

However, if we think about the general requirement of being able to write logic based around the "reduce" operation, and not specifically about how to implement it _in the Terraform language_, there is actually already an answer available today: my "third-party" (that is: this is not an official HashiCorp project) [`apparentlymart/javascript` provider](https://registry.terraform.io/providers/apparentlymart/javascript/latest).

That provider has a JavaScript interpreter embedded in it, which means that if you need to do a computation that is beyond the capabilities of the Terraform language but feasible in the core JavaScript language (without any special external libraries) you can potentially use its `javascript` data source as a solution of last resort.

I actually made that particular provider also embed Underscore.js as a generally-useful library for manipulating collections with functional-style primitives, and so there's [a `reduce` function](https://underscorejs.org/#reduce) built right into it.

Unless we can see some evidence that a general, low-level primitive like "reduce" would be commonly-used and intuitive to many Terraform users, my instinct is to say that this sort of provider-based solution is a reasonable solution of last resort for the rare case when such a thing is needed.

It would also be possible to use a provider to meet a higher-level use-case of for a "reduce"-like operation is an implementation detail. "multiple, sequential regex replacements on a string which may or may not overlap" could be a data source itself, rather than exposing the raw details of how that works into the Terraform language which is usually for higher-level "wiring together" transformations, rather than generalized computation. Perhaps that's what [the `util_replace` data source in `poseidon/util`](https://registry.terraform.io/providers/poseidon/util/latest/docs/data-sources/replace) is doing, though I'm not sure.

**Comment by skyzyx** [comment]
@apparentlymart: I very much appreciate you taking the time to break down the problem and explain all of this.

* So, doing this kind of thing at the HCL level sounds like a no-go, syntax-wise. OK.

* I wasn't aware of the [`apparentlymart/javascript`](https://registry.terraform.io/providers/apparentlymart/javascript/latest) and [`poseidon/util`](https://registry.terraform.io/providers/poseidon/util/latest/docs/data-sources/replace) projects. I've opened them in tabs and will take a look.

> You also asked about whether it could be implemented in a provider.

~~When I mentioned using a provider, I didn't necessarily mean the _expression parsing_ would be done in the provider, rather that _reduce functionality_ might be done inside a provider.~~

It just occurred to me that there isn't a good way to provide a general-purpose reducer without being able to provide the callback function which does the work on each iteration — which is basically `eval`. You can provide multiple task-specific reducers by writing the callback logic in Go on the provider side, though.

Meh.

I'll take a look at these repositories, and maybe try my hand at a provider which contains pre-defined functionality in the form of `resource` blocks.

Thank you again @apparentlymart, as well as @nl-brett-stime for making this GitHub message pop up in my email this morning.

**Comment by crw** [comment]
Thank you for your continued interest in this issue. 

[Terraform version 1.8](https://github.com/hashicorp/terraform/blob/v1.8.0-beta1/CHANGELOG.md) launches with support of provider-defined functions. It is now possible to implement your own functions! We would love to see this implemented as a provider-defined function. 

Please see the [provider-defined functions documentation](https://developer.hashicorp.com/terraform/plugin/framework/functions) to learn how to implement functions in your providers. If you are new to provider development, learn how to [create a new provider](https://developer.hashicorp.com/terraform/plugin/framework) with the Terraform Plugin Framework. If you have any questions, please visit the [Terraform Plugin Development category in our official forum](https://discuss.hashicorp.com/c/terraform-providers/tf-plugin-sdk/43). 

We hope this feature unblocks future function development and provides more flexibility for the Terraform community. Thank you for your continued support of Terraform!

**Comment by FredericGervais** [comment]
If you want a MapReduce() in pure terraform code, you will need to create a sequential for loop.

For loops in terraform run in parallel, but there is a way to create a sequential one.

Like I mentionned here : [https://stackoverflow.com/a/79674158/23894622](https://stackoverflow.com/a/79674158/23894622)

the solution is to use the function :
`templatestring()`

You can base your template on this code that iterates over `x = local.x + 1` 3 times and the output is : 3

```
join("", flatten([
"$${jsonencode(",
"{ for local in {\"\"={x=0}}:\"output\" =>",
"{ for local in {\"\"=merge(local, {x=local.x + 1})}:\"output\" =>",
"{ for local in {\"\"=merge(local, {x=local.x + 1})}:\"output\" =>",
"{ for local in {\"\"=merge(local, {x=local.x + 1})}:\"output\" =>",
"local",
"}.output",
"}.output",
"}.output",
"}.output",
")}"
]))
```

You can then modify the code with classic [list] for loops to generate at run time a sequential for loop that iterates a given amount of time.

Modify the `{x=local.x + 1}` section to have it do the fibonacci sequence of whatever MapReduce operation you need.
